### PR TITLE
[configurator] extracted rendundant functionalities into cmake_var

### PR DESCRIPTION
**Pull request by michelececcacci**
Referencing #114

### CODE REVIEW

File: scripts/configurator/parser/cmake_var.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      "BOOL" : cmake_var_bool,
+      "INTERNAL" : cmake_var,
+      "FILEPATH": cmake_var,
+      "STRING" : cmake_var,
+   }
+   variable_instance = string_to_variable_class[type]
+   return variable_instance(var)
+
+
+class cmake_var:
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_var.py:13]: This should be something like `cmake_var_str`.

File: scripts/configurator/parser/cmake_var.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+   "generic type of cmake_var, used when we don't need particular conversions"
+   def __init__(self, var: str) -> None:
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_var.py:15]: `var` should be `val`. The variable has a _value_ so we're now dealing with it.

File: scripts/configurator/parser/cmake_var.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+def generate_cmake_var(type :str, var):
+   "factory function that associates the correct type of variable"
+   string_to_variable_class = {
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_var.py:3]: Rename to `cmake_type_str_to_class`

File: scripts/configurator/parser/cmake_var.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      "BOOL" : cmake_var_bool,
+      "INTERNAL" : cmake_var,
+      "FILEPATH": cmake_var,
+      "STRING" : cmake_var,
+   }
+   variable_instance = string_to_variable_class[type]
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_var.py:9]: NIT: No need for a variable (too long), but also, `variable_instance` is wrong: you're getting a `class` object, not an instance. So, at most it should be something like `class_obj`. Later, you're creating an object instance from the class.

**Review by vvaltchev — CHANGES_REQUESTED** [review]
Good change! There are a few mostly cosmetic changes to make.
Btw, you didn't use inheritance and classic polymorphism. Now the cmake_var_* classes are completely unrelated. That's totally fine in a dynamic language like Python, but it won't work in other languages like Java. Still, again, no need to forcibly use inheritance here, I'm happy with the current code. I just wanted to point out this aspect.

File: scripts/configurator/parser/cmake_var.py (hunk 1 above)
Comment (michelececcacci) [reply scripts/configurator/parser/cmake_var.py:13]: in place of `var`, right?

File: scripts/configurator/parser/cmake_var.py (hunk 2 above)
Comment (michelececcacci) [reply scripts/configurator/parser/cmake_var.py:15]: Ok, but since when we get the  val  from `cmake_row` it can effectively be either the value (in case of comments) or the `cmake_var`  holding it, i think it's just better to insert a setter to get the value no matter what.

File: scripts/configurator/parser/cmake_var.py (hunk 1 above)
Comment (vvaltchev) [reply scripts/configurator/parser/cmake_var.py:15]: `cmake_var` should be renamed to `cmake_var_str` or, if you want a generic no-op type, you could name it `cmake_var_generic`.

File: scripts/configurator/parser/cmake_var.py (hunk 2 above)
Comment (vvaltchev) [reply scripts/configurator/parser/cmake_var.py]: No, wait. I've been talking about the `var` parameter and `var` member in `cmake_var`. Not the "_var" suffix in the same. The variable (aka cmake_var) has a value. You can call it also `value` in order everything to be even clearer.

File: scripts/configurator/parser/cmake_var.py (hunk 1 above)
Comment (michelececcacci) [reply scripts/configurator/parser/cmake_var.py:15]: Ok. I think switching names between union type and class name is more appropriate

File: scripts/configurator/parser/cmake_var.py (hunk 2 above)
Comment (michelececcacci) [reply scripts/configurator/parser/cmake_var.py]: Ok, makes sense. What about the getter idea? do you like it?  I think makes sense

File: scripts/configurator/parser/cmake_var.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      "BOOL" : cmake_var_bool,
+      "INTERNAL" : cmake_var_generic,
+      "FILEPATH": cmake_var_generic,
+      "STRING" : cmake_var_generic,
+   }
+   class_obj = cmake_type_str_to_class[type]
+   return class_obj(cmake_var_str)
+
+
+class cmake_var_generic:
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_var.py:15]: Sorry for the confusion, in case I haven't been clear enough. I meant to use `cmake_var_generic` if you didn't want to have a base class + inheritance, because that's why I saw in the code and I told you it was fine, even if it wouldn't work on other languages.

Now that you made `cmake_var_bool` to inherit from `cmake_var_generic`, the name `cmake_var_generic` is not good enough. It has to be `cmake_var` like you previously named it or `cmake_var_base` at most. 

In conclusion, either you call it `cmake_var_generic` but don't use inheritance, or you call it `cmake_var` and you do use inheritance. The previous change was a mix (`cmake_var` but without inheritance) and was semantically weird.

File: scripts/configurator/parser/cmake_var.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      "FALSE": False,
+   }
+
+   def __init__(self, value: str) -> None:
+      self.value = self.NORMALIZED_BOOLS[value]
+
+   def serialize(self) -> str:
+      return "ON" if self.value else "OFF"
+
+cmake_var = cmake_var_generic | cmake_var_bool
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_var.py:40]: Can you explain to be how this syntax works? It doesn't look like a linter thing. It looks like a type alternative that I saw only in pure functional languages like Haskell. I didn't know you could do things like that in Python.

Btw, if you use inheritance, you don't use type alternatives like that. You just use the base class, which should be simply called `cmake_var`. That's the whole point of polymorphism. Type alternatives are good when the types are NOT related. It's either one or the other. It doesn't make sense like that.

File: scripts/configurator/parser/cmake_var.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
+from typing import Dict, Any
+
+def generate_cmake_var(type :str, cmake_var_str):
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_var.py:3]: NAMING: This is a factory function, not a generator of any kind. Please don't use the `generate_` prefix. Typically `make` or `build` are used in the industry.

**Review by vvaltchev — CHANGES_REQUESTED** [review]
Very good! There are just a few additional renamings to do related with the previous discussions and one new.

File: scripts/configurator/parser/cmake_var.py (hunk 6 above)
Comment (michelececcacci) [reply scripts/configurator/parser/cmake_var.py:40]: Ok, that makes sense. `X | Y` is just syntactic sugar for `Union[X, Y]`. I think it's also used in typescript.

File: scripts/configurator/parser/cmake_row.py [hunk 8]
```diff
@@ -87,18 +62,16 @@ def is_pound_comment(self, row: str) -> bool:
...
-      serialized_val: str
       if self.row_type == row_type.VARIABLE:
-         if self.configurator_type == configurator_type.BOOL:
-            serialized_val = "ON" if self.val else "OFF"
-         else:
-            serialized_val = str(self.val)
-         return self.name + ":" + self._cmake_type_str + "="  + serialized_val
-      elif self.row_type == row_type.POUND_COMMENT:
+         return self.name + ":" + self._cmake_type + "=" + self.val.serialize()
+      elif type(self.val) == str and self.row_type == row_type.POUND_COMMENT:
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_row.py:67]: No need to check for `type(self.val) == str` here. This is an invariant of your code and always must hold when the row_type is a comment. Therefore, just check for `self.row_type == row_type.POUND_COMMENT` and, if you want, just `assert(type(self.val) == str)` inside the `if`. With `if` statements we should check for things that are expected to be true or false. Invariants are checked with asserts, instead.

Also, the whole serialize() should be conceptually implemented with a big switch statement. Because python doesn't have that, a sequence of if/elif will do the job.

File: scripts/configurator/parser/cmake_row.py [hunk 9]
```diff
@@ -87,18 +62,16 @@ def is_pound_comment(self, row: str) -> bool:
...
          return "#" + self.val
-      elif self.row_type == row_type.SLASH_COMMENT:
+      elif type(self.val) == str and self.row_type == row_type.SLASH_COMMENT:
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_row.py:69]: Same as before.

NIT: It would be more polished to have an `elif` case for the empty string and a final `else` statement where you throw an exception. In other words, the `if/elif` list here must be _exhaustive_ and cover all the types that `row` could possibly have. In the current implementation, you simply assume that there are no more types to cover and default to empty string case: that is correct (obviously) in this case, but "doesn't scale" in large projects with many types to cover. The cool thing about C and C++ switch statements is that, if you do it on an enum type, the compiler will show you a warning in case you missed a case. Anyway, this is an optional NIT as we don't have that scale here and `row` won't have many more types.

File: scripts/configurator/parser/cmake_row.py [hunk 10]
```diff
@@ -87,18 +62,16 @@ def is_pound_comment(self, row: str) -> bool:
...
          return "//" + self.val
       return "" # in case of empty string
 
-   def normalize_bool(self, val: str) -> bool:
-      return self.NORMALIZED_BOOLS[val]
+   def get_val(self):
```
Comment (vvaltchev) [inline scripts/configurator/parser/cmake_row.py:73]: I get what you'd like to have with this `get_val()`. I have a bit of mixed feelings, but I tend for keeping it at the moment, as it does a great job for the tests.

**Review by vvaltchev — CHANGES_REQUESTED** [review]
Almost perfect, very few cosmetic comments.

File: scripts/configurator/parser/cmake_row.py (hunk 8 above)
Comment (michelececcacci) [reply scripts/configurator/parser/cmake_row.py:67]: This is just needed for mypy to stop complaining. The invariant holds, but unfortunately mypy doesn't like it.

File: scripts/configurator/parser/cmake_row.py (hunk 9 above)
Comment (michelececcacci) [reply scripts/configurator/parser/cmake_row.py:69]: Ok, this makes sense

File: scripts/configurator/parser/cmake_row.py (hunk 10 above)
Comment (michelececcacci) [reply scripts/configurator/parser/cmake_row.py:73]: Ok, perfect

File: scripts/configurator/parser/cmake_row.py (hunk 8 above)
Comment (vvaltchev) [reply scripts/configurator/parser/cmake_row.py:67]: OK, then use `str(self.val)` instead of this ugly checking.
